File: frontend/src/lib/permissions.ts
import type { User, UserRole } from '../types';

/** Page keys — must match backend `permissions.ALL_PAGE_KEYS`. */
export type PageKey =
  | 'dashboard'
  | 'visits'
  | 'vehicles'
  | 'fleet'
  | 'services'
  | 'analytics'
  | 'visionflow'
  | 'visionflow_multicam'
  | 'users'
  | 'audit'
  | 'settings';

export const PAGE_LABELS: Record<PageKey, string> = {
  dashboard: 'Dashboard',
  visits: 'Work orders / Visits',
  vehicles: 'Vehicles',
  fleet: 'Fleet intelligence',
  services: 'Services catalog',
  analytics: 'Analytics',
  visionflow: 'ANPR & speed',
  visionflow_multicam: 'Camera wall',
  users: 'Team users',
  audit: 'Audit log',
  settings: 'Settings',
};

export const DEFAULT_PAGES_BY_ROLE: Record<UserRole, PageKey[]> = {
  admin: Object.keys(PAGE_LABELS) as PageKey[],
  manager: (Object.keys(PAGE_LABELS) as PageKey[]).filter(k => k !== 'settings'),
  staff: ['visits'],
  viewer: ['dashboard', 'visits', 'vehicles'],
};

export interface NavItem {
  to: string;
  label: string;
  pageKey: PageKey;
  end?: boolean;
  adminOnly?: boolean;
  adminRoleOnly?: boolean;
}

export const NAV_MAIN: NavItem[] = [
  { to: '/', label: 'Dashboard', pageKey: 'dashboard', end: true },
  { to: '/visits', label: 'Visits', pageKey: 'visits' },
  { to: '/vehicles', label: 'Vehicles', pageKey: 'vehicles' },
  { to: '/fleet-intelligence', label: 'Fleet intelligence', pageKey: 'fleet' },
  { to: '/services', label: 'Services', pageKey: 'services' },
  { to: '/analytics', label: 'Analytics', pageKey: 'analytics' },
  { to: '/visionflow', label: 'ANPR & speed', pageKey: 'visionflow' },
  { to: '/visionflow/multicam', label: 'Camera wall', pageKey: 'visionflow_multicam' },
];

export const NAV_ADMIN: NavItem[] = [
  { to: '/users', label: 'Users', pageKey: 'users', adminOnly: true },
  { to: '/audit', label: 'Audit Log', pageKey: 'audit', adminOnly: true },
  { to: '/settings', label: 'Settings', pageKey: 'settings', adminRoleOnly: true },
];

/** Longest-prefix match for nested routes (e.g. /visits/new → visits). */
const ROUTE_PAGE_MAP: { prefix: string; pageKey: PageKey }[] = [
  { prefix: '/visits', pageKey: 'visits' },
  { prefix: '/vehicles', pageKey: 'vehicles' },
  { prefix: '/fleet-intelligence', pageKey: 'fleet' },
  { prefix: '/services', pageKey: 'services' },
  { prefix: '/analytics', pageKey: 'analytics' },
  { prefix: '/visionflow/multicam', pageKey: 'visionflow_multicam' },
  { prefix: '/visionflow/history', pageKey: 'visionflow' },
  { prefix: '/visionflow', pageKey: 'visionflow' },
  { prefix: '/users', pageKey: 'users' },
  { prefix: '/audit', pageKey: 'audit' },
  { prefix: '/settings', pageKey: 'settings' },
  { prefix: '/', pageKey: 'dashboard' },
];

export function effectivePages(user: User | null | undefined): PageKey[] {
  if (!user) return [];
  if (user.allowed_pages?.length) return user.allowed_pages as PageKey[];
  return DEFAULT_PAGES_BY_ROLE[user.role] ?? ['visits'];
}

export function canAccessPage(user: User | null | undefined, pageKey: PageKey): boolean {
  return effectivePages(user).includes(pageKey);
}

export function pageKeyForPath(pathname: string): PageKey {
  const path = pathname.split('?')[0];
  for (const { prefix, pageKey } of ROUTE_PAGE_MAP) {
    if (prefix === '/') {
      if (path === '/') return pageKey;
      continue;
    }
    if (path === prefix || path.startsWith(`${prefix}/`)) return pageKey;
  }
  return 'dashboard';
}

export function canAccessRoute(user: User | null | undefined, pathname: string): boolean {
  const key = pageKeyForPath(pathname);
  if (!canAccessPage(user, key)) return false;
  if (key === 'settings' && user?.role !== 'admin') return false;
  if ((key === 'users' || key === 'audit') && user && !['admin', 'manager'].includes(user.role)) {
    return canAccessPage(user, key);
  }
  return true;
}

export function getHomeRoute(user: User | null | undefined): string {
  const pages = effectivePages(user);
  if (pages.includes('dashboard')) return '/';
  if (pages.includes('visits')) return '/visits';
  const first = NAV_MAIN.find(n => pages.includes(n.pageKey));
  return first?.to ?? '/visits';
}

export function isOrgWideUser(user: User | null | undefined): boolean {
  return user?.role === 'admin' || user?.role === 'manager';
}

export function filterNavItems(items: NavItem[], user: User | null | undefined): NavItem[] {
  const pages = effectivePages(user);
  return items.filter(item => {
    if (item.adminRoleOnly && user?.role !== 'admin') return false;
    if (item.adminOnly && user && !['admin', 'manager'].includes(user.role) && !pages.includes(item.pageKey)) {
      return false;
    }
    return pages.includes(item.pageKey);
  });
}

File: backend/app/services/permissions.py
"""Role defaults and per-user page permissions."""

from __future__ import annotations

import json
from typing import TYPE_CHECKING

from sqlalchemy import or_

from ..models.user import UserRole

if TYPE_CHECKING:
    from sqlalchemy.orm import Query
    from ..models.user import User
    from ..models.visit import Visit

# Keys exposed to the frontend route guard
ALL_PAGE_KEYS: tuple[str, ...] = (
    "dashboard",
    "visits",
    "vehicles",
    "fleet",
    "services",
    "analytics",
    "visionflow",
    "visionflow_multicam",
    "users",
    "audit",
    "settings",
)

PAGE_LABELS: dict[str, str] = {
    "dashboard": "Dashboard",
    "visits": "Work orders / Visits",
    "vehicles": "Vehicles",
    "fleet": "Fleet intelligence",
    "services": "Services catalog",
    "analytics": "Analytics",
    "visionflow": "ANPR & speed",
    "visionflow_multicam": "Camera wall",
    "users": "Team users",
    "audit": "Audit log",
    "settings": "Settings",
}

DEFAULT_PAGES_BY_ROLE: dict[UserRole, list[str]] = {
    UserRole.ADMIN: list(ALL_PAGE_KEYS),
    UserRole.MANAGER: [k for k in ALL_PAGE_KEYS if k != "settings"],
    UserRole.STAFF: ["visits"],
    UserRole.VIEWER: ["dashboard", "visits", "vehicles"],
}


def parse_stored_pages(raw: str | None) -> list[str] | None:
    if raw is None or not str(raw).strip():
        return None
    try:
        data = json.loads(raw)
        if isinstance(data, list):
            cleaned = [str(p) for p in data if str(p) in ALL_PAGE_KEYS]
            return cleaned
    except (json.JSONDecodeError, TypeError):
        pass
    return None


def dump_stored_pages(pages: list[str] | None) -> str | None:
    if pages is None:
        return None
    cleaned = [p for p in pages if p in ALL_PAGE_KEYS]
    return json.dumps(cleaned)


def effective_pages(user: "User") -> list[str]:
    custom = parse_stored_pages(getattr(user, "allowed_pages", None))
    if custom is not None:
        return custom
    role = user.role if isinstance(user.role, UserRole) else UserRole(str(user.role))
    return list(DEFAULT_PAGES_BY_ROLE.get(role, ["visits"]))


def has_org_wide_access(user: "User") -> bool:
    role = user.role if isinstance(user.role, UserRole) else UserRole(str(user.role))
    return role in (UserRole.ADMIN, UserRole.MANAGER)


def can_access_page(user: "User", page_key: str) -> bool:
    return page_key in effective_pages(user)


def user_owns_visit(user: "User", visit: "Visit") -> bool:
    """Staff may only see visits they created or supervisor-signed."""
    if has_org_wide_access(user):
        return True
    uid = user.id
    return visit.created_by == uid or visit.supervisor_signed_by == uid


def apply_visit_scope(q: "Query", user: "User"):
    if has_org_wide_access(user):
        return q
    return q.filter(
        or_(
            Visit.created_by == user.id,
            Visit.supervisor_signed_by == user.id,
        )
    )


# Late import for Visit in apply_visit_scope
from ..models.visit import Visit  # noqa: E402
